Feature/topping suggestion

jaalazard opened the pull request:
## Description

Please include a summary of the change or which issue is fixed.

## Type of change

- [x ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Area of change

- [ ] Frontend
- [ x] Backend
- [ ] Other: .....

## General checklist:

- [ x] My code follows the style guidelines of this project
- [ x] I ran `bun run format`/`bun run lint` before commit
- [ x] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] I have made corresponding changes to the documentation where needed
- [ x] My changes generate no new warnings
- [ x] I have checked my code and corrected any misspellings
- [ x] I have updated `develop` and merged to my branch before submitting pull request
- [ x] My pull request generate no conflicts with `develop` branch
- [ x] I requested code review from other team members

## Frontend checklist:

- [ ] I followed guidelines for `HTML`/`JSX`/`Typescript`
- [ ] My `Javascript` generate no new console errors
- [ ] I tested my code cross browsers
- [ ] My slice is pixel perfect for both desktop and mobile according to design
- [ ] I conducted basic QA to assure all features are working
- [ ] I tested responsive for mobile and tablet resolutions

## Backend checklist:

- [x ] I tested admin by manually adding content from zero
- [ x] I created easy to use admin experience which is self-explanatory
- [ x] I added description to admin fields in hard-to-understand areas
- [ x] I followed guidelines for naming `Typescript` variables
- [ x] I conducted basic QA to assure all features are working

jaalazard committed:
feat: generate a random topping

jaalazard committed:
feat: generating a topping list based on alcohol & ingredient selection

Justicesvr commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
intéressant de faire un système de points

jaalazard committed:
feat: generating a topping list based on alcohol & ingredient selection

YDrogen commented on packages/backend/api/src/router.ts:
```diff
@@ -43,6 +44,16 @@ router.get('/alcohols/:level', async (req, res) => {
+router.get('/toppings/:mainFlavour', async (req, res) => {
+  const mainFlavour: Flavour = req.params.mainFlavour as Flavour;
+  try {
+    const result = await getToppingsByFlavour(db, mainFlavour);
```
Pourquoi envoyer `db` dans la fonction ?

jaalazard replied:
Parce que getToppingsByFlavour attend deux arguments : la db, et la mainFlavour

YDrogen replied:
Pourquoi `getToppingsByFlavour` attend l'argument `db` ?

jaalazard replied:
Parce que je l'ai défini ainsi, c'est un problème ?

YDrogen replied:
D'où sort le `db` ici ?

jaalazard replied:
de @app/backend-shared

YDrogen commented on packages/backend/api/src/router.ts:
```diff
@@ -43,6 +44,16 @@ router.get('/alcohols/:level', async (req, res) => {
+  } catch {
+    res.status(500).json({ error: 'Internal Server Error' });
+  }
+});
```
Ça serait sympa de mettre ça dans un fichier dédié comme pour le reste

YDrogen commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
```diff
@@ -1,17 +1,106 @@
+        break;
+      }
+      case 'spicy': {
+        flavoursCount.spicy += 1;
+        break;
+      }
+      case 'herbaceous': {
+        flavoursCount.herbaceous += 1;
+        break;
+      }
+      case 'floral': {
+        flavoursCount.floral += 1;
+        break;
+      }
+      case 'woody': {
+        flavoursCount.woody += 1;
+        break;
+      }
+      case 'bitter': {
+        flavoursCount.bitter += 1;
+        break;
+      }
+      case 'sweet': {
+        flavoursCount.sweet += 1;
+        break;
+      }
+      case 'salty': {
+        flavoursCount.salty += 1;
+        break;
+      }
+      case 'sour': {
+        flavoursCount.sour += 1;
+        break;
+      }
+      case 'neutral': {
+        flavoursCount.neutral += 1;
+        break;
+      }
+    }
+  }
```
Tu peux transformer tout ça en 1 reduce
Avec un genre de `flavoursCount = [selectedAlcohol?.flavour, selectedIngredient?.flavour].reduce()`

jaalazard replied:
C'était l'idée mais : `Array#reduce()` is not allowedeslint[unicorn/no-array-reduce](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v49.0.0/docs/rules/no-array-reduce.md)

YDrogen replied:
Tu crois aux licornes toi ?
Tu peux ajouter `"unicorn/no-array-reduce": "off",` dans le fichier `.eslintrc.cjs` (à placer dans l'ordre alphabétique) avec les autres règles

jaalazard replied:
Je crois aux yummyCorns. Ah bon je ne pensais pas qu'on pouvais péter tout ton paramétrage, mais avec plaisir !

YDrogen replied:
Je vous autorise à rajouter cette ligne 👍

YDrogen commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
```diff
@@ -1,17 +1,106 @@
+    }
+  }
+
+  let mainFlavour = '';
+  let mainFlavourCount = 0;
+
+  for (const [key, count] of Object.entries(flavoursCount)) {
+    if (count > mainFlavourCount) {
+      mainFlavour = key;
+      mainFlavourCount = count;
+    }
+  }
```
Les `let` sont interdits dans les components comme ça !
Il faut trouver une façon de faire avec des `const`

YDrogen reviewed (changes requested):
Où est l'algo de suggestion ? Qu'est-il censé faire ?

jaalazard commented:
> Où est l'algo de suggestion ? Qu'est-il censé faire ?

Dans Toppingpart.tsx, il calcule les points de "flavour" et suggère un topping associé à la flavour gagnante.

jaalazard committed:
feat: generating a topping list based on alcohol & ingredient selection

YDrogen commented:
> > Où est l'algo de suggestion ? Qu'est-il censé faire ?
> 
> Dans Toppingpart.tsx, il calcule les points de "flavour" et suggère un topping associé à la flavour gagnante.

Mais là le calcul est fait avec 1 alcool et 1 ingrédient, non ?
Et si la flavour ne correspond à aucun topping, rien d'autre n'est proposé ?
Et si il n'y a pas de flavour, il n'y a pas de topping proposé ?

jaalazard commented:
> > > Où est l'algo de suggestion ? Qu'est-il censé faire ?
> > 
> > 
> > Dans Toppingpart.tsx, il calcule les points de "flavour" et suggère un topping associé à la flavour gagnante.
> 
> Mais là le calcul est fait avec 1 alcool et 1 ingrédient, non ? Et si la flavour ne correspond à aucun topping, rien d'autre n'est proposé ? Et si il n'y a pas de flavour, il n'y a pas de topping proposé ?

Absolument, en attendant d'avoir le choix de l'ingrédient 2 et de l'ingrédient 3 qui viendront ajouter des flavour au calcul. Chaque alcohol et chaque ingrédient a une flavour donc j'obtiendrai forcément une flavour à la fin. Et nous aurons à minima un topping par flavour, à nous d'alimenter la bdd

YDrogen commented:
> > > > Où est l'algo de suggestion ? Qu'est-il censé faire ?
> > > 
> > > 
> > > Dans Toppingpart.tsx, il calcule les points de "flavour" et suggère un topping associé à la flavour gagnante.
> > 
> > 
> > Mais là le calcul est fait avec 1 alcool et 1 ingrédient, non ? Et si la flavour ne correspond à aucun topping, rien d'autre n'est proposé ? Et si il n'y a pas de flavour, il n'y a pas de topping proposé ?
> 
> Absolument, en attendant d'avoir le choix de l'ingrédient 2 et de l'ingrédient 3 qui viendront ajouter des flavour au calcul. Chaque alcohol et chaque ingrédient a une flavour donc j'obtiendrai forcément une flavour à la fin. Et nous aurons à minima un topping par flavour, à nous d'alimenter la bdd

Je vois. 🤔 
Du coup c'est pas possible d'avoir de suggestions qui ne match pas avec notre cocktail histoire d'avoir des combos incroyables ? :(

jaalazard commented:
> > > > > Où est l'algo de suggestion ? Qu'est-il censé faire ?
> > > > 
> > > > 
> > > > Dans Toppingpart.tsx, il calcule les points de "flavour" et suggère un topping associé à la flavour gagnante.
> > > 
> > > 
> > > Mais là le calcul est fait avec 1 alcool et 1 ingrédient, non ? Et si la flavour ne correspond à aucun topping, rien d'autre n'est proposé ? Et si il n'y a pas de flavour, il n'y a pas de topping proposé ?
> > 
> > 
> > Absolument, en attendant d'avoir le choix de l'ingrédient 2 et de l'ingrédient 3 qui viendront ajouter des flavour au calcul. Chaque alcohol et chaque ingrédient a une flavour donc j'obtiendrai forcément une flavour à la fin. Et nous aurons à minima un topping par flavour, à nous d'alimenter la bdd
> 
> Je vois. 🤔 Du coup c'est pas possible d'avoir de suggestions qui ne match pas avec notre cocktail histoire d'avoir des combos incroyables ? :(

Tout est possible après... Je peux suggérer un topping cohérent et 3 topping qui n'ont rien à voir, ou suggérer des topping dont la flavour est l'opposé de celle des ingrédients... Compliqué... Quel est le contraire de fruité si ce n'est "pas fruité" ? Quel est le contraire de "terreux" ? Qui a tué Kennedy ?

Sinon on peut ajouter l'option random/"tête de mort" comme pour les ingrédients genre "à vos risques et périls" ?

jaalazard committed:
feat: generating a topping list based on alcohol & ingredient selection

jaalazard committed:
feat: generating a topping list based on alcohol & ingredient selection

YDrogen commented:
> > > > > > Où est l'algo de suggestion ? Qu'est-il censé faire ?
> > > > > 
> > > > > 
> > > > > Dans Toppingpart.tsx, il calcule les points de "flavour" et suggère un topping associé à la flavour gagnante.
> > > > 
> > > > 
> > > > Mais là le calcul est fait avec 1 alcool et 1 ingrédient, non ? Et si la flavour ne correspond à aucun topping, rien d'autre n'est proposé ? Et si il n'y a pas de flavour, il n'y a pas de topping proposé ?
> > > 
> > > 
> > > Absolument, en attendant d'avoir le choix de l'ingrédient 2 et de l'ingrédient 3 qui viendront ajouter des flavour au calcul. Chaque alcohol et chaque ingrédient a une flavour donc j'obtiendrai forcément une flavour à la fin. Et nous aurons à minima un topping par flavour, à nous d'alimenter la bdd
> > 
> > 
> > Je vois. 🤔 Du coup c'est pas possible d'avoir de suggestions qui ne match pas avec notre cocktail histoire d'avoir des combos incroyables ? :(
> 
> Tout est possible après... Je peux suggérer un topping cohérent et 3 topping qui n'ont rien à voir, ou suggérer des topping dont la flavour est l'opposé de celle des ingrédients... Compliqué... Quel est le contraire de fruité si ce n'est "pas fruité" ? Quel est le contraire de "terreux" ? Qui a tué Kennedy ?
> 
> Sinon on peut ajouter l'option random/"tête de mort" comme pour les ingrédients genre "à vos risques et périls" ?

Tu peux en prio proposer quelque chose qui match, et après faire du random yes

jaalazard committed:
feat: generating a topping list based on alcohol & ingredient selection

jaalazard committed:
ok

Merge branch 'develop' of github.com:WildCodeSchool/2023-09-orleans-cda-projet2-groupe-a into feature/topping-suggestion

jaalazard committed:
feat: generating a topping list based on alcohol & ingredient selection

Justicesvr commented on packages/backend/api/src/services/topping-service.ts:
magnifique random

YDrogen commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
```diff
@@ -1,17 +1,109 @@
+    const allFlavours = [selectedAlcohol?.flavour, selectedIngredient?.flavour];
+
+    // eslint-disable-next-line unicorn/no-array-reduce
+    return allFlavours.reduce<Record<string, number>>(
```
Tu peux pas mettre 
```suggestion
    return allFlavours.reduce<Record<Flavour, number>>(
```
ici ? Tu pourrais avoir de l'autocomplétion grâce à ça

YDrogen commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
```diff
@@ -1,17 +1,109 @@
+  };
+
+  useEffect(() => {
+    if (isRandomToppingChoosen && randomTopping) {
+      setIsShowRandomTopping(true);
```
Ça peut être une `const` dans le component ça
`const shouldBlabla = isRandomTruc && randomTruc`

YDrogen commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
```diff
@@ -1,17 +1,109 @@
+  const [isShowRandomTopping, setIsShowRandomTopping] =
+    useState<boolean>(false);
+
+  const memorizedFlavoursCount = useMemo(() => {
```
Faute d'orthographe
```suggestion
  const memoizedFlavoursCount = useMemo(() => {
```

YDrogen commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
```diff
@@ -29,34 +121,50 @@ export default function ToppingPart({
-          </div>
-        ))}
-      </fieldset>
+      {isShowRandomTopping && randomTopping ? (
```
`randomTopping` sert à rien, c'est forcément défini si `isShowRandomTopping` est `true`

jaalazard committed:
feat: generate a random topping : changes done

YDrogen commented on packages/frontend/web/src/components/form-cocktail/ToppingPart.tsx:
```diff
@@ -121,8 +116,8 @@ export default function ToppingPart({
       ) : undefined}
-      {isShowRandomTopping && randomTopping ? (
-        <p>{randomTopping.name}</p>
+      {isShowRandomTopping ? (
```
Comme dans le useEffect au dessus, tu peux remplacer la condition par la const que tu as créé
Tu n'auras plus besoin du ? en dessous normalement

jaalazard committed:
feat: generate a random topping : changes done

jaalazard committed:
feat: generate a random topping : changes done
